vault/token_store.go

```diff
@@ -558,51 +572,96 @@ func (ts *TokenStore) storeCommon(entry *TokenEntry, writeSecondary bool) error
 	return nil
 }
 
-// UseToken is used to manage restricted use tokens and decrement
-// their available uses. Note: this is potentially racy, but the simple
-// solution of a global lock would be severely detrimental to performance. Also
-// note the specific revoke case below.
-func (ts *TokenStore) UseToken(te *TokenEntry) error {
-	// If the token is not restricted, there is nothing to do
-	if te.NumUses == 0 {
-		return nil
+func (ts *TokenStore) getTokenLock(id string) *sync.RWMutex {
+	// Find our multilevel lock, or fall back to global
+	var lock *sync.RWMutex
+	var ok bool
+	if len(id) >= 2 {
+		lock, ok = ts.tokenLocks[id[0:2]]
+	}
+	if !ok || lock == nil {
+		lock = ts.tokenLocks["global"]
```

Comment:
`global` handles the cases where the token ID has been specified by a root user. We can't really validate those ahead of time; in order to do the lookup we should already have a read lock. This is likely a very minimal set of tokens but we should still lock.